src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,181 @@
+            tail = head;
+            size++;
+            return;
+        }
+        Node<T> currentNode = new Node<>(tail, value, null);
+        tail.next = currentNode;
+        tail = currentNode;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("Index is not valid");
+        } else {
+            /* if we add Node to the end*/
+            if (index == size) {
+                add(value);
+                return;
+            }
+            /* if we add Node to the beginning */
+            if (index == 0) {
+                Node<T> newNode = new Node<>(null, value, head);
+                head.prev = newNode;
+                head = newNode;
+                size++;
+                return;
+            }
+            Node<T> currentNode = getNodeByIndex(index);
+            Node<T> newNode = new Node<>(currentNode.prev, value, currentNode);
+            currentNode.prev.next = newNode;
+            currentNode.prev = newNode;
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        indexIsValid(index);
+        return getNodeByIndex(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        indexIsValid(index);
+        T temp = getNodeByIndex(index).item;
+        getNodeByIndex(index).item = value;
+        return temp;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        indexIsValid(index);
+        /* if we need remove single Node */
+        if (size == 1 && index == 0) {
+            size--;
+            Node<T> removedNode = tail;
+            tail = head = null;
+            return removedNode.item;
+        }
+        /* if we need remove Node from the end */
+        if (index == size - 1) {
+            size--;
+            Node<T> removedNode = tail;
+            tail = tail.prev;
+            tail.next.prev = tail.next = null;
+            return removedNode.item;
+        }
+        /* if we need remove Node from the beginning*/
+        if (index == 0) {
+            size--;
+            Node<T> removedNode = head;
+            head = head.next;
+            head.prev.next = head.prev = null;
+            return removedNode.item;
+        }
+        Node<T> currentNode = getNodeByIndex(index);
+        currentNode.prev.next = currentNode.next;
+        currentNode.next.prev = currentNode.prev;
+        size--;
+        return currentNode.item;
     }
 
     @Override
     public boolean remove(T object) {
-        return false;
+        if (!isValid(getIndexByValue(object))) {
+            return false;
+        }
+        remove(getIndexByValue(object));
```

Comment:
You move in list two times. If not good for big list. You can create local variable.